include/nlohmann/detail/meta/type_traits.hpp

```diff
@@ -150,6 +149,52 @@ struct has_to_json < BasicJsonType, T, enable_if_t < !is_basic_json<T>::value >>
 ///////////////////
 // is_ functions //
 ///////////////////
+// https://en.cppreference.com/w/cpp/types/conjunction
+template<class...> struct conjunction : std::true_type { };
+template<class B1> struct conjunction<B1> : B1 { };
+template<class B1, class... Bn>
+struct conjunction<B1, Bn...>
+: std::conditional<bool(B1::value), conjunction<Bn...>, B1>::type {};
+
+// Reimplementation of is_constructible and is_default_constructible, due to them being broken for
+// std::pair and std::tuple until LWG 2367 fix (see https://cplusplus.github.io/LWG/lwg-defects.html#2367).
+// This causes compile errors in e.g. clang 3.5 or gcc 4.9.
+// Based on commit fixing this in gcc: https://github.com/gcc-mirror/gcc/commit/d3c64041b32b6962ad6b2d879231537a477631fb
```

Comment:
Can you elaborate whether you took code out of that commit? I'm asking because of licensing issues (GPLv2 vs. MIT).

Reply:
I looked at the code to see what the "proper" fix looked like. However, that requires modifying `std::pair`/`std::tuple`. The code I added is more or less the same as the suggested one in the LWG defect report. Well, the logic is the same, it's not the exact same code. So I don't think there's any issue here.

It's also this report that mentions that this fix is incomplete:
> The proposed resolution is incomplete, because it wouldn't work for cv-qualified objects of pair or for references of them during reference-initialization.

I don't think this is an issue, since `detail::is_constructible` is never called for e.g. `std::pair<A, B> volatile` nor for a reference to it. I did not add extra specializations to `detail::is_constructible` for `volatile` types or references. Currently only `std::pair<A, B>` and `std::pair<A, B> const` are specialized. Actually, I don't think that second one will ever be used. Do you prefer I add specializations for `volatile`, `const volatile` and references to all specialized types anyway (i.e. 6 extra specializations for `std::pair` and `std::tuple` each)?

Reply:
This is not what I meant. I was asking whether the code following that comment was written by yourself (and could be MIT licensed in this project) or taken from a different source - e.g., the GCC source code which would be GPLv2 licensed.

Reply:
No no, as I mentioned, the code I added uses the same logic as the suggested solution mentioned in the LWG defect report. No code was copied from anywhere.

Reply:
Yes, please. Comments like this make FOSS auditors anxious...

Reply:
Something to keep in mind next time :smile:.